client/service/src/builder.rs

```diff
@@ -371,41 +368,79 @@ pub fn new_client<E, Block, RA>(
-	/// An optional finality proof request builder.
-	pub finality_proof_request_builder: Option<BoxFinalityProofRequestBuilder<TBl>>,
-	/// An optional, shared finality proof request provider.
-	pub finality_proof_provider: Option<Arc<dyn FinalityProofProvider<TBl>>>,
 	/// A shared transaction pool.
 	pub transaction_pool: Arc<TExPool>,
 	/// A RPC extension builder. Use `NoopRpcExtensionBuilder` if you just want to pass in the
 	/// extensions directly.
 	pub rpc_extensions_builder: Box<dyn RpcExtensionBuilder<Output = TRpc> + Send>,
 	/// An optional, shared remote blockchain instance. Used for light clients.
 	pub remote_blockchain: Option<Arc<dyn RemoteBlockchain<TBl>>>,
-	/// A block annouce validator builder.
-	pub block_announce_validator_builder:
-		Option<Box<dyn FnOnce(Arc<TCl>) -> Box<dyn BlockAnnounceValidator<TBl> + Send> + Send>>,
+	/// A shared network instance.
+	pub network: Arc<NetworkService<TBl, <TBl as BlockT>::Hash>>,
+	/// Sinks to propagate network status updates.
+	pub network_status_sinks: NetworkStatusSinks<TBl>,
+	/// A Sender for RPC requests.
+	pub system_rpc_tx: TracingUnboundedSender<sc_rpc::system::Request<TBl>>,
+	/// Shared Telemetry connection sinks,
+	pub telemetry_connection_sinks: TelemetryConnectionSinks,
+}
+
+/// Build a shared offchain workers instance.
+pub fn build_offchain_workers<TBl, TBackend, TCl>(
+	config: &Configuration,
+	backend: Arc<TBackend>, spawn_handle: SpawnTaskHandle, client: Arc<TCl>,
+	network: Arc<NetworkService<TBl, <TBl as BlockT>::Hash>>,
+) -> Option<Arc<sc_offchain::OffchainWorkers<TCl, TBackend::OffchainStorage, TBl>>>
+	where
+		TBl: BlockT, TBackend: sc_client_api::Backend<TBl>,
+		<TBackend as sc_client_api::Backend<TBl>>::OffchainStorage: 'static,
+		TCl: Send + Sync + ProvideRuntimeApi<TBl> + BlockchainEvents<TBl> + 'static,
+		<TCl as ProvideRuntimeApi<TBl>>::Api: sc_offchain::OffchainWorkerApi<TBl>,
+{
+	let offchain_workers = match backend.offchain_storage() {
+		Some(db) => {
+			Some(Arc::new(sc_offchain::OffchainWorkers::new(client.clone(), db)))
+		},
+		None => {
+			warn!("Offchain workers disabled, due to lack of offchain storage support in backend.");
+			None
+		},
+	};
+
+	// Inform the offchain worker about new imported blocks
+	if let Some(offchain) = offchain_workers.clone() {
+		spawn_handle.spawn(
+			"offchain-notifications",
+			sc_offchain::notification_future(
+				config.role.is_authority(),
+				client.clone(),
+				offchain,
+				Clone::clone(&spawn_handle),
+				network.clone()
+			)
+		);
+	}
+
+	offchain_workers
 }
 
 /// Put together the components of a service from the parameters.
-pub fn build<TBl, TBackend, TImpQu, TExPool, TRpc, TCl>(
-	builder: ServiceParams<TBl, TCl, TImpQu, TExPool, TRpc, TBackend>,
-) -> Result<ServiceComponents<TBl, TBackend, TCl>, Error>
+pub fn spawn_tasks<TBl, TBackend, TExPool, TRpc, TCl>(
+	params: SpawnTasksParams<TBl, TCl, TExPool, TRpc, TBackend>,
+) -> Result<Arc<RpcHandlers>, Error>
```

Comment:
I wish that this could just return a `Result<(), Error>`, but a lot of params are shared between the non-rpc parts of this function and `gen_handler`.